chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala

```diff
@@ -570,23 +575,36 @@ class Bundle(implicit compileOptions: CompileOptions) extends Record {
 
     // Check if this is an inner class, and if so, try to get the outer instance
     val clazz = this.getClass
+
     val outerClassInstance = Option(clazz.getEnclosingClass).map { outerClass =>
       def canAssignOuterClass(x: Object) = outerClass.isAssignableFrom(x.getClass)
-      val outerInstance = try {
-        clazz.getDeclaredField("$outer").get(this)  // doesn't work in all cases, namely anonymous inner Bundles
-      } catch {
-        case _: NoSuchFieldException =>
-          // First check if this Bundle is bound and an anonymous inner Bundle of another Bundle
-          this.bindingOpt.collect { case ChildBinding(p) if canAssignOuterClass(p) => p }
-              .orElse(this._outerInst)
-              .getOrElse(reflectError(s"Unable to determine instance of outer class $outerClass"))
-      }
-      if (!canAssignOuterClass(outerInstance)) {
-        reflectError(s"Unable to determine instance of outer class $outerClass," +
-            s" guessed $outerInstance, but is not assignable to outer class $outerClass")
+
+      val outerInstance = _outerInst match {
+        case Some(outerInstance) => outerInstance  // use _outerInst if defined
+        case None =>  // determine outer instance if not already recorded
+          try {
+            // Prefer this if it works, but doesn't work in all cases, namely anonymous inner Bundles
+            val outer = clazz.getDeclaredField("$outer").get(this)
+            _outerInst = Some(outer)
+            outer
+          } catch {
+            case _: NoSuchFieldException =>
```

Comment:
I have been unable to produce a small example (still working on it), but I have run into `IllegalAccessException`s on cases that this PR otherwise handles, eg.
`java.lang.IllegalAccessException: Class chisel3.core.Bundle$$anonfun$11 can not access a member of class mypackage.MyModule$$anon$2 with modifiers "private final"`

Note that the code that caused this issue was merely anonymous Bundles nested inside `Decoupled` and modifying the catch block to catch `IllegalAccessException`s fixed the problem and `autoclonetype` worked just fine.

Any ideas why this might occur?

Reply:
Played around with trying to reproduce this, was only able to get the error to come up when accessing `$outer` from outside the Bundle (in the testdriver class) object. So it seems like class-based access restrictions are being enforced even with reflection.

That being said, Bundle should only be reflecting on itself, which makes this really odd. I'm also surprised that autoclonetype is being called from some `class chisel3.core.Bundle$$anonfun$11` rather than the bundle object `class mypackage.MyModule$$anon$2` itself. Maybe some edge case with how Scala translates things down to Java?